waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/SaslThriftMetastoreClientManager.java

```diff
@@ -0,0 +1,233 @@
+  }
+
+  @Override
+  void open(HiveUgiArgs ugiArgs) {
+    if (isConnected) {
+      return;
+    }
+    createMetastoreClientAndOpen(null, ugiArgs);
+    if (impersonationEnabled) {
+      try {
+        String userName = UserGroupInformation.getCurrentUser().getShortUserName();
+        DelegationTokenKey key = new DelegationTokenKey(msUri, userName, client);
+        String delegationToken = delegationTokenCache.get(key);
+        close();
+        createMetastoreClientAndOpen(delegationToken, ugiArgs);
+      } catch (IOException | ExecutionException e) {
+        log.error("Couldn't create delegation token client");
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  void createMetastoreClientAndOpen(String delegationToken, HiveUgiArgs ugiArgs) {
+    TException te = null;
+    boolean useSsl = conf.getBoolVar(ConfVars.HIVE_METASTORE_USE_SSL);
+    boolean useCompactProtocol = conf.getBoolVar(ConfVars.METASTORE_USE_THRIFT_COMPACT_PROTOCOL);
+    int clientSocketTimeout = (int) conf.getTimeVar(ConfVars.METASTORE_CLIENT_SOCKET_TIMEOUT,
+        TimeUnit.MILLISECONDS);
+
+    for (int attempt = 0; !isConnected && (attempt < retries); ++attempt) {
+      for (URI store : metastoreUris) {
+        log.info("Trying to connect to metastore with URI {}", store);
+        try {
+          transport = new TSocket(store.getHost(), store.getPort(), clientSocketTimeout,
+              connectionTimeout);
+          // Wrap thrift connection with SASL for secure connection.
+          try {
+            UserGroupInformation.setConfiguration(conf);
+
+            // check if we should use delegation tokens to authenticate
+            // the call below gets hold of the tokens if they are set up by hadoop
+            // this should happen on the map/reduce tasks if the client added the
+            // tokens into hadoop's credential store in the front end during job
+            // submission.
+            if (impersonationEnabled && delegationToken != null) {
+              // authenticate using delegation tokens via the "DIGEST" mechanism
+              transport = KerberosSaslHelper
+                  .getTokenTransport(delegationToken,
+                      store.getHost(), transport,
+                      MetaStoreUtils.getMetaStoreSaslProperties(conf, useSsl));
+            } else {
+              String principalConfig = conf.getVar(ConfVars.METASTORE_KERBEROS_PRINCIPAL);
+              transport = UserGroupInformation.getLoginUser().doAs(
+                  (PrivilegedExceptionAction<TTransport>) () -> KerberosSaslHelper.getKerberosTransport(
+                      principalConfig, store.getHost(), transport,
+                      MetaStoreUtils.getMetaStoreSaslProperties(conf, useSsl), false));
+            }
+          } catch (IOException | InterruptedException exception) {
+            log.error("Couldn't create client transport, URI " + store, exception);
```

Comment:
If you're wrapping the exception, include the message in the log statement in the exception's message. Then, remove the logging statement because it's redundant.

Reply:
This log will print the exception stack, which is very helpful for solving problem. The below `throw new MetaException(exception.toString());` will loss exception stack, because MetaException cannot init via `MetaException(IOException | InterruptedException)`,`MetaException(String) ` will loss exception stack. From this perspective ，I think it is not redundant.